// gonzales-pe/lib/node-types.js
module.exports = {
    ArgumentsType: 'arguments',
    AtkeywordType: 'atkeyword',
    AtrulebType: 'atruleb',
    AtrulerType: 'atruler',
    AtrulerqType: 'atrulerq',
    AtrulersType: 'atrulers',
    AtrulesType: 'atrules',
    AttribType: 'attribute',
    AttrselectorType: 'attributeSelector',
    BlockType: 'block',
    BracesType: 'braces',
    ClassType: 'class',
    CombinatorType: 'combinator',
    CommentMLType: 'multilineComment',
    CommentSLType: 'singlelineComment',
    ConditionType: 'condition',
    ConditionalStatementType: 'conditionalStatement',
    DeclarationType: 'declaration',
    DeclDelimType: 'declarationDelimiter',
    DefaultType: 'default',
    DelimType: 'delimiter',
    DimensionType: 'dimension',
    EscapedStringType: 'escapedString',
    FilterType: 'filter',
    FiltervType: 'filterValue',
    FunctionType: 'function',
    FunctionExpressionType: 'functionExpression',
    IdentType: 'ident',
    ImportantType: 'important',
    IncludeType: 'include',
    InterpolationType: 'interpolation',
    InterpolatedVariableType: 'interpolatedVariable',
    LoopType: 'loop',
    MixinType: 'mixin',
    NamespaceType: 'namespace',
    NthType: 'nth',
    NthselectorType: 'nthSelector',
    NumberType: 'number',
    OperatorType: 'operator',
    ParentSelectorType: 'parentSelector',
    PercentageType: 'percentage',
    PlaceholderType: 'placeholder',
    ProgidType: 'progid',
    PropertyType: 'property',
    PropertyDelimType: 'propertyDelimiter',
    PseudocType: 'pseudoClass',
    PseudoeType: 'pseudoElement',
    RawType: 'raw',
    RulesetType: 'ruleset',
    SType: 'space',
    SelectorType: 'selector',
    ShashType: 'id',
    SimpleselectorType: 'simpleSelector',
    StringType: 'string',
    StylesheetType: 'stylesheet',
    UnaryType: 'unaryOperator',
    UriType: 'uri',
    ValueType: 'value',
    VariableType: 'variable',
    VariablesListType: 'variablesList',
    VhashType: 'color'
};
